### PR TITLE
[AMDGPU] Fix machine verification failure from INIT_EXEC lowering

### PR DESCRIPTION
**Pull request by jayfoad**
Fix machine verification failure from INIT_EXEC lowering since it was
moved from SILowerControlFlow to SIWholeQuadMode in #94452.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp [hunk 1]
```diff
@@ -1676,6 +1676,8 @@ bool SIWholeQuadMode::runOnMachineFunction(MachineFunction &MF) {
   if (!(GlobalFlags & (StateWQM | StateStrict)) && LowerToCopyInstrs.empty() &&
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1676]: I am now even more convinced that we should not bother with this fast path. We should just make sure that the general path is fast when there is no work to do.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp [hunk 2]
```diff
@@ -1717,7 +1719,7 @@ bool SIWholeQuadMode::runOnMachineFunction(MachineFunction &MF) {
   LIS->removeAllRegUnitsForPhysReg(AMDGPU::SCC);
 
   // If we performed any kills then recompute EXEC
-  if (!KillInstrs.empty())
+  if (!KillInstrs.empty() || !InitExecInstrs.empty())
     LIS->removeAllRegUnitsForPhysReg(AMDGPU::EXEC);
```
Comment (arsenm) [inline llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1723]: This shouldn't be necessary in the first place. Reserved registers, like exec, do not have liveness tracked

Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1723]: Well I see stuff like this in the dumps:
```
# After SI Whole Quad Mode
********** INTERVALS **********
EXEC_LO_LO16 [16r,16d:0) 0@16r
EXEC_LO_HI16 [16r,16d:0) 0@16r
EXEC_HI_LO16 [16r,16d:0) 0@16r
EXEC_HI_HI16 [16r,16d:0) 0@16r
```

Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1723]: Those shouldn't have been created in the first place. Where did they come from?

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 1 above)
Comment (perlfu) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1676]: I'll run compile time testing on a version without it -- it's possible there really is little difference already.

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 2 above)
Comment (perlfu) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1723]: In this test case they are formed in Register Coalescer when it attempts to merge a copy of EXEC with a virtual and asks for an interval on a EXEC reg unit.

Registered registers still have intervals, they are a sequence of a dead defs, one for each assignment to the register.
You can see this in `LiveIntervals::computeRegUnitRange`.
I do not know the use for this, but assume it is for interference calculations.

**Review by perlfu — APPROVED** [review]
LGTM

File: llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp (hunk 1 above)
Comment (jayfoad) [reply llvm/lib/Target/AMDGPU/SIWholeQuadMode.cpp:1676]: The only thing I can see that we _might_ be able to skip over in some cases is the two `for (auto BII : Blocks)` loops below.

**Pull request merged**